File: api-gateway/app/models/response.py
from typing import Optional, Any, Generic, TypeVar
from pydantic import BaseModel

T = TypeVar("T")

class PaginationMeta(BaseModel):
    total: int = 0
    limit: int = 0
    page: int = 0
    total_pages: int = 0
    has_next: bool = False
    has_previous: bool = False

class APIResponse(BaseModel, Generic[T]):
    success: bool
    data: Optional[T] = None
    error: Optional[str] = None
    message: str
    meta: PaginationMeta = PaginationMeta()

File: user-service/src/utils/response.ts
export interface PaginationMeta {
  total: number
  limit: number
  page: number
  total_pages: number
  has_next: boolean
  has_previous: boolean
}

export interface ApiResponse<T> {
  success: boolean
  data?: T
  error?: string
  message: string
  meta?: PaginationMeta
}

// Helper functions for consistent formatting
export const successResponse = <T>(
  data: T,
  message = "Success",
  meta?: PaginationMeta
): ApiResponse<T> => ({
  success: true,
  data,
  message,
  ...(meta ? { meta } : {}),
});

export const errorResponse = (
  message: string,
  error?: string
): ApiResponse<null> => ({
  success: false,
  message,
  ...(error ? { error } : {}),
});
